tests/pull.rs

```diff
@@ -0,0 +1,218 @@
+    let query = r#"[:find ?hood (pull ?district [:district/name :district/region])
+                    :where
+                    (or [?hood :neighborhood/name "Beacon Hill"]
+                        [?hood :neighborhood/name "Capitol Hill"])
+                    [?hood :neighborhood/district ?district]
+                    :order ?hood]"#;
+    let results: RelResult<Binding> = store.begin_read().expect("read")
+                                           .q_once(query, None)
+                                           .into_rel_result()
+                                           .expect("results");
+
+    let beacon_district: Vec<(NamespacedKeyword, TypedValue)> = vec![
+        (kw!(:district/name), "Greater Duwamish".into()),
+        (kw!(:district/region), schema.get_entid(&NamespacedKeyword::new("region", "se")).unwrap().into())
+    ];
+    let beacon_district: StructuredMap = beacon_district.into();
+    let capitol_district: Vec<(NamespacedKeyword, TypedValue)> = vec![
+        (kw!(:district/name), "East".into()),
+        (kw!(:district/region), schema.get_entid(&NamespacedKeyword::new("region", "e")).unwrap().into())
+    ];
+    let capitol_district: StructuredMap = capitol_district.into();
+
+    let expected = RelResult {
+                       width: 2,
+                       values: vec![
+                           TypedValue::Ref(capitol).into(), capitol_district.into(),
+                           TypedValue::Ref(beacon).into(), beacon_district.into(),
+                       ].into(),
+                   };
+    assert_eq!(results, expected.clone());
+
+    // We can also prepare the query.
+    let reader = store.begin_read().expect("read");
+    let mut prepared = reader.q_prepare(query, None).expect("prepared");
+
+    assert_eq!(prepared.run(None)
+                       .into_rel_result()
+                       .expect("results"),
+               expected);
+
+    // Execute a scalar query where the body is constant.
+    // TODO: we shouldn't require `:where`; that makes this non-constant!
+    let query = r#"[:find (pull ?hood [:neighborhood/name]) . :in ?hood
+                    :where [?hood :neighborhood/district _]]"#;
+    let result = reader.q_once(query, QueryInputs::with_value_sequence(vec![(var!(?hood), TypedValue::Ref(beacon))]))
+                       .into_scalar_result()
+                       .expect("success")
+                       .expect("result");
+
+    let expected: StructuredMap = vec![(kw!(:neighborhood/name), TypedValue::from("Beacon Hill"))].into();
+    assert_eq!(result, expected.into());
+
+    // Collect the names and regions of all districts.
+    let query = r#"[:find [(pull ?district [:district/name :district/region]) ...]
+                    :where [_ :neighborhood/district ?district]]"#;
+    let results = reader.q_once(query, None)
+                        .into_coll_result()
+                        .expect("result");
+
+    let expected: Vec<StructuredMap> = vec![
```

Comment:
Quick, nested maps so that we can `(pull ?d [{:district/region [:region/name]}])` (or whatever the syntax is) so these store-specific entids don't appear.

Reply:
It's on my list!